Location.py

```diff
@@ -71,63 +88,89 @@ def add_rule(self, lambda_rule):
-        return (self.parent_region.can_fill(item, manual) and self.item_rule(self, item))
-
-
-    def is_disabled(self):
-        return (self.disabled == DisableType.DISABLED) or \
-               (self.disabled == DisableType.PENDING and self.locked)
-
+    @property
+    def is_disabled(self) -> bool:
+        return ((self.disabled == DisableType.DISABLED) or
+                (self.disabled == DisableType.PENDING and self.locked))
 
     # Can the player see what's placed at this location without collecting it?
     # Used to reduce JSON spoiler noise
-    def has_preview(self):
+    def has_preview(self) -> bool:
+        if self.world is None:
+            return False
         return location_is_viewable(self.name, self.world.settings.correct_chest_appearances, self.world.settings.fast_chests)
 
-
-    def has_item(self):
+    def has_item(self) -> bool:
         return self.item is not None
 
-    def has_no_item(self):
+    def has_no_item(self) -> bool:
         return self.item is None
 
-    def has_progression_item(self):
+    def has_progression_item(self) -> bool:
         return self.item is not None and self.item.advancement
 
-
-    def __str__(self):
+    def maybe_set_misc_hints(self) -> None:
+        if self.item is None or self.item.world is None or self.world is None:
+            return
+        if self.item.world.dungeon_rewards_hinted and self.item.name in self.item.world.rewardlist:
+            if self.item.name not in self.item.world.hinted_dungeon_reward_locations:
+                self.item.world.hinted_dungeon_reward_locations[self.item.name] = self
+                logging.getLogger('').debug(f'{self.item.name} [{self.item.world.id}] set to [{self.name}]')
+        for hint_type in misc_item_hint_table:
+            item = self.item.world.misc_hint_items[hint_type]
+            if hint_type not in self.item.world.misc_hint_item_locations and self.item.name == item:
+                self.item.world.misc_hint_item_locations[hint_type] = self
+                logging.getLogger('').debug(f'{item} [{self.item.world.id}] set to [{self.name}]')
+        for hint_type in misc_location_hint_table:
+            the_location = self.world.misc_hint_locations[hint_type]
+            if hint_type not in self.world.misc_hint_location_items and self.name == the_location:
+                self.world.misc_hint_location_items[hint_type] = self.item
+                logging.getLogger('').debug(f'{the_location} [{self.world.id}] set to [{self.item.name}]')
+
+    def __str__(self) -> str:
         return str(self.__unicode__())
 
-
-    def __unicode__(self):
+    def __unicode__(self) -> str:
         return '%s' % self.name
 
 
-def LocationFactory(locations, world=None):
+@overload
+def LocationFactory(locations: str) -> Location:
+    pass
+
+
+@overload
+def LocationFactory(locations: list[str]) -> list[Location]:
+    pass
+
+
+def LocationFactory(locations: str | list[str]) -> Location | list[Location]:
```

Comment:
Since python doesn't have real overloaded functions, the `@overload` versions of it above just serve as type hints. In this case, it lets your IDE/whatever type parser know that if you send it a single string, you get back a single location, and if you pass it a list of strings, you get back a list of locations. Without the `@overload`s, each return is assumed to be either a location or a list of locations - it doesn't know which one would have been returned.

Reply:
Huh, alright